// @sumup-oss/circuit-ui/dist/components/Table/components/TableBody/TableBody.js
"use strict";
"use client";
import { jsx } from "react/jsx-runtime";
import { mapRowProps, mapCellProps } from "../../utils.js";
import { TableRow } from "../TableRow/TableRow.js";
import { TableHeader } from "../TableHeader/TableHeader.js";
import { TableCell } from "../TableCell/TableCell.js";
function TableBody({
  rows = [],
  condensed,
  rowHeaders = false,
  sortHover,
  onRowClick
}) {
  return /* @__PURE__ */ jsx("tbody", { children: rows.map((row, rowIndex) => {
    const { cells, ...props } = mapRowProps(row);
    return /* @__PURE__ */ jsx(
      TableRow,
      {
        onClick: onRowClick ? () => onRowClick(rowIndex) : void 0,
        ...props,
        children: cells.map(
          (cell, cellIndex) => rowHeaders && cellIndex === 0 ? /* @__PURE__ */ jsx(
            TableHeader,
            {
              fixed: true,
              condensed,
              scope: "row",
              isHovered: sortHover === cellIndex,
              sortParams: { sortable: false },
              ...mapCellProps(cell)
            },
            `table-cell-${rowIndex}-${cellIndex}`
          ) : /* @__PURE__ */ jsx(
            TableCell,
            {
              condensed,
              isHovered: sortHover === cellIndex,
              ...mapCellProps(cell)
            },
            `table-cell-${rowIndex}-${cellIndex}`
          )
        )
      },
      `table-row-${rowIndex}`
    );
  }) });
}
export {
  TableBody
};
