// @sumup-oss/circuit-ui/dist/components/ComparisonTable/components/BooleanValue/BooleanValue.js
"use strict";
import { jsxs, jsx } from "react/jsx-runtime";
import { utilClasses } from "../../../../styles/utility.js";
import classes from "./BooleanValue.module.css.js";
const BooleanValue = ({ label, value }) => /* @__PURE__ */ jsxs("div", { className: classes.base, children: [
  /* @__PURE__ */ jsx("span", { className: utilClasses.hideVisually, children: label }),
  value ? /* @__PURE__ */ jsxs(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: "24",
      height: "24",
      viewBox: "0 0 24 24",
      fill: "none",
      "aria-hidden": "true",
      "data-testid": "boolean-value-true",
      children: [
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M0.000244141 12C0.000244141 5.37258 5.37283 0 12.0002 0C18.6277 0 24.0002 5.37258 24.0002 12C24.0002 18.6274 18.6277 24 12.0002 24C5.37283 24 0.000244141 18.6274 0.000244141 12Z",
            fill: "var(--cui-bg-success)"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M10.0002 16.4L6.00024 12.4L7.40024 11L10.0002 13.6L16.6002 7L18.0002 8.4L10.0002 16.4Z",
            fill: "var(--cui-fg-success)"
          }
        )
      ]
    }
  ) : /* @__PURE__ */ jsx(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: "24",
      height: "24",
      viewBox: "0 0 24 24",
      fill: "none",
      "aria-hidden": "true",
      "data-testid": "boolean-value-false",
      children: /* @__PURE__ */ jsx(
        "rect",
        {
          x: "6",
          y: "11",
          width: "12",
          height: "2",
          fill: "var(--cui-fg-placeholder)"
        }
      )
    }
  )
] });
export {
  BooleanValue
};
