// @sumup-oss/circuit-ui/dist/components/Carousel/components/Buttons/Buttons.js
"use strict";
import { jsx } from "react/jsx-runtime";
import { ChevronLeft, ChevronRight, Play, Pause } from "@sumup-oss/icons";
import { clsx } from "../../../../styles/clsx.js";
import classes from "./Buttons.module.css.js";
import { IconButton } from "../../../Button/IconButton.js";
const ButtonList = ({ className, ...props }) => /* @__PURE__ */ jsx("div", { className: clsx(classes["button-list"], className), ...props });
const NextButton = (props) => /* @__PURE__ */ jsx(
  IconButton,
  {
    ...props,
    className: clsx(classes.button, props.className),
    size: "s",
    icon: ChevronRight
  }
);
const PrevButton = (props) => /* @__PURE__ */ jsx(
  IconButton,
  {
    ...props,
    className: clsx(classes.button, props.className),
    size: "s",
    icon: ChevronLeft
  }
);
const PlayButton = ({
  paused,
  ...props
}) => /* @__PURE__ */ jsx(
  IconButton,
  {
    ...props,
    className: clsx(classes.button, props.className),
    size: "s",
    icon: paused ? Play : Pause
  }
);
export {
  ButtonList,
  NextButton,
  PlayButton,
  PrevButton
};
