// @sumup-oss/circuit-ui/dist/components/ProgressBar/ProgressBar.js
"use strict";
import { jsxs, jsx } from "react/jsx-runtime";
import { useId } from "react";
import { isSufficientlyLabelled, AccessibilityError } from "../../util/errors.js";
import { utilClasses } from "../../styles/utility.js";
import { clsx } from "../../styles/clsx.js";
import { deprecate } from "../../util/logger.js";
import classes from "./ProgressBar.module.css.js";
function getWidth(value = 0, max = 1) {
  const width = value && max ? (value / max * 100).toFixed(2) : 0;
  return `${width}%`;
}
const legacySizeMap = {
  byte: "s",
  kilo: "m",
  mega: "l"
};
function ProgressBar({
  max,
  value,
  size: legacySize = "m",
  duration = 3e3,
  loop = false,
  paused = false,
  label,
  hideLabel,
  className,
  ...props
}) {
  const ariaId = useId();
  if (process.env.NODE_ENV !== "production" && process.env.NODE_ENV !== "test" && !isSufficientlyLabelled(label)) {
    throw new AccessibilityError(
      "ProgressBar",
      "The `label` prop is missing or invalid."
    );
  }
  if (process.env.NODE_ENV !== "production" && legacySizeMap[legacySize]) {
    deprecate(
      "ProgressBar",
      `The \`${legacySize}\` size has been deprecated. Use the \`${legacySizeMap[legacySize]}\` size instead.`
    );
  }
  const size = legacySizeMap[legacySize] || legacySize;
  return /* @__PURE__ */ jsxs("div", { className: clsx(classes.wrapper, className), ...props, children: [
    max || value ? (
      // biome-ignore lint/a11y/useFocusableInteractive: Progress bars aren't focusable elements, but they can be accessed with a screen reader's read/browse mode.
      /* @__PURE__ */ jsx(
        "span",
        {
          role: "progressbar",
          "aria-valuenow": value,
          "aria-valuemin": 0,
          "aria-valuemax": max,
          "aria-labelledby": ariaId,
          className: clsx(classes.base, classes[size]),
          style: { "--pagination-width": getWidth(value, max) }
        }
      )
    ) : (
      // biome-ignore lint/a11y/useFocusableInteractive: Progress bars aren't focusable elements, but they can be accessed with a screen reader's read/browse mode.
      /* @__PURE__ */ jsx(
        "span",
        {
          role: "progressbar",
          "aria-labelledby": ariaId,
          className: clsx(classes.base, classes[size]),
          "data-loop": loop,
          style: {
            "--pagination-animation-duration": `${duration}ms`,
            "--pagination-animation-play-state": paused ? "paused" : "running"
          }
        }
      )
    ),
    /* @__PURE__ */ jsx(
      "span",
      {
        id: ariaId,
        className: clsx(classes.label, hideLabel && utilClasses.hideVisually),
        children: label
      }
    )
  ] });
}
export {
  ProgressBar
};
