// @sumup-oss/circuit-ui/dist/components/SidePanel/components/Header/Header.js
"use strict";
"use client";
import { jsxs, jsx } from "react/jsx-runtime";
import { ArrowLeft } from "@sumup-oss/icons";
import { clsx } from "../../../../styles/clsx.js";
import classes from "./Header.module.css.js";
import { IconButton } from "../../../Button/IconButton.js";
import { Headline } from "../../../Headline/Headline.js";
import { CloseButton } from "../../../CloseButton/CloseButton.js";
const Header = ({
  backButtonLabel,
  closeButtonLabel,
  headline,
  id,
  isSticky,
  onBack,
  onClose
}) => /* @__PURE__ */ jsxs("div", { className: clsx(classes.base, isSticky && classes.sticky), children: [
  onBack && backButtonLabel && /* @__PURE__ */ jsx(
    IconButton,
    {
      className: classes.button,
      type: "button",
      onClick: onBack,
      icon: ArrowLeft,
      variant: "tertiary",
      children: backButtonLabel
    }
  ),
  /* @__PURE__ */ jsx(Headline, { id, size: "s", as: "h2", className: classes.headline, children: headline }),
  closeButtonLabel && /* @__PURE__ */ jsx(CloseButton, { className: classes.button, onClick: onClose, children: closeButtonLabel })
] });
export {
  Header
};
