// @sumup-oss/circuit-ui/dist/components/Compact/Compact.module.css.js
const base = "cui-compact-srlg";
const l = "cui-compact-l-jb6s";
const m = "cui-compact-m-edk5";
const s = "cui-compact-s-6ovr";
const regular = "cui-compact-regular-wsvc";
const semibold = "cui-compact-semibold-yqdn";
const bold = "cui-compact-bold-8esh";
const normal = "cui-compact-normal-kot0";
const subtle = "cui-compact-subtle-ekxa";
const placeholder = "cui-compact-placeholder-wct2";
const accent = "cui-compact-accent-ify9";
const success = "cui-compact-success-ehhn";
const warning = "cui-compact-warning-ou5t";
const danger = "cui-compact-danger-9m0k";
const promo = "cui-compact-promo-gbjy";
const classes = {
  base,
  l,
  m,
  s,
  regular,
  semibold,
  bold,
  normal,
  subtle,
  placeholder,
  "on-strong": "cui-compact-on-strong-lcgn",
  "on-strong-subtle": "cui-compact-on-strong-subtle-dyop",
  accent,
  success,
  warning,
  danger,
  promo
};
export {
  accent,
  base,
  bold,
  danger,
  classes as default,
  l,
  m,
  normal,
  placeholder,
  promo,
  regular,
  s,
  semibold,
  subtle,
  success,
  warning
};
